File: Dart/14-Backtracking/Algorithms/3-HamiltonianCycle/HamiltonianCycle.dart
import 'dart:math';
import 'package:collection/collection.dart'; // Import the collection package

// 1. Backtracking Approach
bool findHamiltonianCycleBacktracking(List<List<int>> graph, List<int> path, int pos) {
  int n = graph.length;

  // Base case: all vertices are included in the path
  if (pos == n) {
    // Check if there is an edge from the last vertex to the starting vertex
    if (graph[path[pos - 1]][path[0]] == 1) {
      return true;
    } else {
      return false;
    }
  }

  // Try different vertices as the next vertex in the path
  for (int v = 0; v < n; v++) {
    // Check if this vertex is adjacent to the previous vertex and is not already in the path
    if (graph[path[pos - 1]][v] == 1 && !path.contains(v)) {
      path[pos] = v;
      // Recursively try to find a Hamiltonian Cycle
      if (findHamiltonianCycleBacktracking(graph, path, pos + 1)) {
        return true;
      }
      // Backtrack: if the recursive call does not find a solution, remove the vertex from the path
      path[pos] = -1;
    }
  }

  // If no solution is found, return false
  return false;
}

List<int>? hamiltonianCycleBacktracking(List<List<int>> graph) {
  int n = graph.length;
  List<int> path = List<int>.filled(n, -1);
  path[0] = 0; // Start from vertex 0

  if (findHamiltonianCycleBacktracking(graph, path, 1)) {
    return path;
  } else {
    return null;
  }
}

// 2. Backtracking with Adjacency Set Optimization
bool findHamiltonianCycleBacktrackingAdjSet(Map<int, Set<int>> adjList, List<int> path, int pos) {
  int n = adjList.length;

  if (pos == n) {
    // Check for edge between last and first vertex.
    return adjList[path[pos - 1]]?.contains(path[0]) ?? false;
  }

  final neighbors = adjList[path[pos - 1]] ?? {}; // Get neighbors, handle null

  for (int v in neighbors) {
    if (!path.contains(v)) {
      path[pos] = v;
      if (findHamiltonianCycleBacktrackingAdjSet(adjList, path, pos + 1)) {
        return true;
      }
      path[pos] = -1; // Backtrack
    }
  }
  return false;
}

List<int>? hamiltonianCycleBacktrackingAdjSet(List<List<int>> graph) {
  int n = graph.length;
  Map<int, Set<int>> adjList = {};
  for (int i = 0; i < n; i++) {
    adjList[i] = {};
    for (int j = 0; j < n; j++) {
      if (graph[i][j] == 1) {
        adjList[i]!.add(j);
      }
    }
  }

  List<int> path = List<int>.filled(n, -1);
  path[0] = 0;

  if (findHamiltonianCycleBacktrackingAdjSet(adjList, path, 1)) {
    return path;
  } else {
    return null;
  }
}

// 3. Optimized Backtracking with Degree Check
bool findHamiltonianCycleOptimized(List<List<int>> graph, List<int> path, int pos, List<int> degrees) {
  int n = graph.length;

  if (pos == n) {
    return graph[path[pos - 1]][path[0]] == 1;
  }

  for (int v = 0; v < n; v++) {
    if (graph[path[pos - 1]][v] == 1 && !path.contains(v)) {
      path[pos] = v;
      if (findHamiltonianCycleOptimized(graph, path, pos + 1, degrees)) {
        return true;
      }
      path[pos] = -1;
    }
  }
  return false;
}

List<int>? hamiltonianCycleOptimized(List<List<int>> graph) {
  int n = graph.length;
  List<int> path = List<int>.filled(n, -1);
  List<int> degrees = List<int>.filled(n, 0);

  // Calculate degrees (optimization: check if any node has degree < 2)
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      if (graph[i][j] == 1) {
        degrees[i]++;
      }
    }
    if (degrees[i] < 2) {
      return null; // No Hamiltonian Cycle if a vertex has degree < 2
    }
  }

  path[0] = 0;
  if (findHamiltonianCycleOptimized(graph, path, 1, degrees)) {
    return path;
  } else {
    return null;
  }
}

// 4.  Warnsdorff's Rule (for Knight's Tour - a special case, but showing a different heuristic)
List<int>? knightTourWarnsdorff(int size) {
  // Only works for Knight's Tour (L-shaped moves on a square grid)
  if (size <= 0) return null;
  List<List<int>> board = List.generate(size, (_) => List<int>.filled(size, -1));
  // Possible knight moves (relative offsets)
  final List<int> dx = [2, 1, -1, -2, -2, -1, 1, 2];
  final List<int> dy = [1, 2, 2, 1, -1, -2, -2, -1];

  bool isValid(int x, int y) {
    return x >= 0 && x < size && y >= 0 && y < size && board[x][y] == -1;
  }

  int getDegree(int x, int y) {
    int count = 0;
    for (int i = 0; i < 8; i++) {
      if (isValid(x + dx[i], y + dy[i])) {
        count++;
      }
    }
    return count;
  }

  // Custom sort function to sort neighbors by their degree.  Important for Warnsdorff's
  List<int> getSortedNeighbors(int x, int y) {
    List<int> neighbors = [];
    List<int> degrees = []; // Store degrees of neighbors
    for (int i = 0; i < 8; i++) {
      int nextX = x + dx[i];
      int nextY = y + dy[i];
      if (isValid(nextX, nextY)) {
        neighbors.add(i); // Store the *index* of the move, not the coord
        degrees.add(getDegree(nextX, nextY));
      }
    }
    // Sort neighbors based on their degrees (Warnsdorff's heuristic)
    List<int> sortedNeighbors = [];
    while(neighbors.isNotEmpty){
      int minDegree = degrees.reduce((a, b) => a < b ? a : b);
      int minIndex = degrees.indexOf(minDegree);
      sortedNeighbors.add(neighbors[minIndex]);
      neighbors.removeAt(minIndex);
      degrees.removeAt(minIndex);
    }
    return sortedNeighbors;
  }

  bool solveKnightTour(int x, int y, int moveCount) {
    board[x][y] = moveCount;
    if (moveCount == size * size - 1) {
      return true; // Tour is complete
    }

    List<int> sortedNeighbors = getSortedNeighbors(x, y); // Get sorted neighbors

    for (int moveIndex in sortedNeighbors) { // Iterate through *sorted* neighbors
      int nextX = x + dx[moveIndex];
      int nextY = y + dy[moveIndex];
      if (solveKnightTour(nextX, nextY, moveCount + 1)) {
        return true;
      }
    }

    board[x][y] = -1; // Backtrack
    return false;
  }

  // Start the tour from the top-left corner (0, 0)
  if (solveKnightTour(0, 0, 0)) {
    // Convert the 2D board to a 1D path (Hamiltonian Cycle representation)
    List<int> path = [];
    for (int i = 0; i < size; i++) {
      for (int j = 0; j < size; j++) {
        // Find the cell with move number 'k' and add its index to the path
        int k = i * size + j;
        for (int r = 0; r < size; r++) {
          for (int c = 0; c < size; c++) {
            if (board[r][c] == k) {
              path.add(r * size + c); // Convert 2D coords to 1D index
              break;
            }
          }
        }
      }
    }
    return path;
  } else {
    return null;
  }
}

// 5. Randomized Approach (for general Hamiltonian Cycle, not very efficient)
List<int>? hamiltonianCycleRandomized(List<List<int>> graph) {
  int n = graph.length;
  List<int> path = List<int>.filled(n, -1);
  List<int> vertices = List<int>.generate(n, (i) => i); // [0, 1, 2, ..., n-1]
  final random = Random();

  for (int startVertex = 0; startVertex < n; startVertex++) { // Try each start vertex
    path[0] = startVertex;
    List<int> remainingVertices = List<int>.from(vertices);
    remainingVertices.remove(startVertex);
    for (int i = 1; i < n; i++) {
      // Shuffle the remaining vertices to get a random order.
      remainingVertices.shuffle(random);
      bool foundNext = false;
      for (int j = 0; j < remainingVertices.length; j++) {
        int nextVertex = remainingVertices[j];
        if (graph[path[i - 1]][nextVertex] == 1) {
          path[i] = nextVertex;
          remainingVertices.removeAt(j);
          foundNext = true;
          break;
        }
      }
      if (!foundNext) {
        // If no valid next vertex is found, backtrack and try a different starting vertex
        path.fillRange(i, n, -1);
        break;
      }
    }

    if (path[n - 1] != -1 && graph[path[n - 1]][path[0]] == 1) {
      return path; // Found a Hamiltonian Cycle
    }
    path.fillRange(1, n, -1); // Clear path, except for the first vertex
  }
  return null; // No Hamiltonian Cycle found
}



void main() {
  // Example graph represented as an adjacency matrix
  List<List<int>> graph = [
    [0, 1, 0, 0, 1],
    [1, 0, 1, 0, 0],
    [0, 1, 0, 1, 0],
    [0, 0, 1, 0, 1],
    [1, 0, 0, 1, 0],
  ];

  print("Hamiltonian Cycle Example:");
  print("Graph (Adjacency Matrix):");
  for (var row in graph) {
    print(row);
  }
  print("\n");

  // 1. Backtracking
  List<int>? path1 = hamiltonianCycleBacktracking(graph);
  print("1. Backtracking:");
  if (path1 != null) {
    print("Hamiltonian Cycle: $path1");
  } else {
    print("No Hamiltonian Cycle found.");
  }

  // 2. Backtracking with Adjacency Set
  List<int>? path2 = hamiltonianCycleBacktrackingAdjSet(graph);
  print("\n2. Backtracking with Adjacency Set:");
  if (path2 != null) {
    print("Hamiltonian Cycle: $path2");
  } else {
    print("No Hamiltonian Cycle found.");
  }

  // 3. Optimized Backtracking
  List<int>? path3 = hamiltonianCycleOptimized(graph);
  print("\n3. Optimized Backtracking:");
  if (path3 != null) {
    print("Hamiltonian Cycle: $path3");
  } else {
    print("No Hamiltonian Cycle found.");
  }

  // 4. Warnsdorff's Rule (Knight's Tour)
  print("\n4. Warnsdorff's Rule (Knight's Tour - Size 8):");
  List<int>? knightTourPath = knightTourWarnsdorff(8); // Example for an 8x8 board
  if (knightTourPath != null) {
    print("Knight's Tour Path: $knightTourPath");
  } else {
    print("No Knight's Tour found.");
  }

  // 5. Randomized Approach
  List<int>? path5 = hamiltonianCycleRandomized(graph);
  print("\n5. Randomized Approach:");
  if (path5 != null) {
    print("Hamiltonian Cycle: $path5");
  } else {
    print("No Hamiltonian Cycle found.");
  }
}

File: Java/14-Backtracking/Algorithms/3-HamiltonianCycle/HamiltonianCycle.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class HamiltonianCycle {

    private static final int V = 5; // Number of vertices in the graph.  Can be changed.

    // 1. Basic Backtracking Approach
    //   - Uses a simple recursive backtracking algorithm.
    //   - Tries all possible paths, and checks if they form a Hamiltonian Cycle.
    //   - Not very efficient for larger graphs.
    public static boolean hasHamiltonianCycle1(int[][] graph) {
        int[] path = new int[V];
        Arrays.fill(path, -1); // Initialize path array with -1
        path[0] = 0; // Start from vertex 0
        return solveHamiltonianCycle1(graph, path, 1);
    }

    private static boolean solveHamiltonianCycle1(int[][] graph, int[] path, int pos) {
        // Base case: All vertices are included in the path
        if (pos == V) {
            // Check if there is an edge from the last vertex to the starting vertex (0)
            if (graph[path[pos - 1]][0] == 1) {
                return true; // Form a cycle
            } else {
                return false; // Path is not a cycle
            }
        }

        // Try different vertices as the next vertex in the path
        for (int v = 0; v < V; v++) {
            // Check if the current vertex v can be added to the path
            if (isSafe1(v, graph, path, pos)) {
                path[pos] = v; // Add vertex v to the path
                // Recursively try to find a path for the remaining vertices
                if (solveHamiltonianCycle1(graph, path, pos + 1)) {
                    return true; // Solution found
                }
                // Backtrack: Remove vertex v from the path and try other options
                path[pos] = -1;
            }
        }
        // If no vertex can be added to the path, return false
        return false;
    }

    private static boolean isSafe1(int v, int[][] graph, int[] path, int pos) {
        // Check if the vertex v is adjacent to the previous vertex in the path
        if (graph[path[pos - 1]][v] == 0) {
            return false; // Not adjacent
        }

        // Check if the vertex v is already included in the path
        for (int i = 0; i < pos; i++) {
            if (path[i] == v) {
                return false; // Already in path
            }
        }
        return true; // Vertex v is safe to add
    }


    // 2. Backtracking with Adjacency List
    //    - Uses an adjacency list to represent the graph, which can be more efficient for sparse graphs.
    public static boolean hasHamiltonianCycle2(List<List<Integer>> adjList) {
        int[] path = new int[V];
        Arrays.fill(path, -1);
        path[0] = 0;
        return solveHamiltonianCycle2(adjList, path, 1);
    }

    private static boolean solveHamiltonianCycle2(List<List<Integer>> adjList, int[] path, int pos) {
        if (pos == V) {
            if (adjList.get(path[pos - 1]).contains(0)) {
                return true;
            } else {
                return false;
            }
        }

        for (int v : adjList.get(path[pos - 1])) { // Iterate over neighbors
            if (isSafe2(v, path, pos)) {
                path[pos] = v;
                if (solveHamiltonianCycle2(adjList, path, pos + 1)) {
                    return true;
                }
                path[pos] = -1;
            }
        }
        return false;
    }

    private static boolean isSafe2(int v, int[] path, int pos) {
        for (int i = 0; i < pos; i++) {
            if (path[i] == v) {
                return false;
            }
        }
        return true;
    }

    // 3. Backtracking with Degree Check
    //    - Before starting the backtracking, it checks if every vertex has a degree of at least 2.
    //    - A Hamiltonian Cycle is not possible if any vertex has a degree less than 2.
    public static boolean hasHamiltonianCycle3(int[][] graph) {
        for (int i = 0; i < V; i++) {
            int degree = 0;
            for (int j = 0; j < V; j++) {
                if (graph[i][j] == 1) {
                    degree++;
                }
            }
            if (degree < 2) {
                return false; // Hamiltonian Cycle not possible
            }
        }
        int[] path = new int[V];
        Arrays.fill(path, -1);
        path[0] = 0;
        return solveHamiltonianCycle3(graph, path, 1);
    }

    private static boolean solveHamiltonianCycle3(int[][] graph, int[] path, int pos) {
        if (pos == V) {
            if (graph[path[pos - 1]][0] == 1) {
                return true;
            } else {
                return false;
            }
        }

        for (int v = 0; v < V; v++) {
            if (isSafe3(v, graph, path, pos)) {
                path[pos] = v;
                if (solveHamiltonianCycle3(graph, path, pos + 1)) {
                    return true;
                }
                path[pos] = -1;
            }
        }
        return false;
    }

    private static boolean isSafe3(int v, int[][] graph, int[] path, int pos) {
        if (graph[path[pos - 1]][v] == 0) {
            return false;
        }
        for (int i = 0; i < pos; i++) {
            if (path[i] == v) {
                return false;
            }
        }
        return true;
    }

    // 4. Backtracking with Forward Checking
    //    - After adding a vertex to the path, it checks if there are enough unvisited neighbors
    //      for the remaining vertices in the path.
    //    - This can help prune the search space and improve efficiency.
    public static boolean hasHamiltonianCycle4(int[][] graph) {
        int[] path = new int[V];
        Arrays.fill(path, -1);
        path[0] = 0;
        return solveHamiltonianCycle4(graph, path, 1);
    }

    private static boolean solveHamiltonianCycle4(int[][] graph, int[] path, int pos) {
        if (pos == V) {
            if (graph[path[pos - 1]][0] == 1) {
                return true;
            } else {
                return false;
            }
        }

        for (int v = 0; v < V; v++) {
            if (isSafe4(v, graph, path, pos)) {
                path[pos] = v;
                if (solveHamiltonianCycle4(graph, path, pos + 1)) {
                    return true;
                }
                path[pos] = -1;
            }
        }
        return false;
    }

    private static boolean isSafe4(int v, int[][] graph, int[] path, int pos) {
        if (graph[path[pos - 1]][v] == 0) {
            return false;
        }
        for (int i = 0; i < pos; i++) {
            if (path[i] == v) {
                return false;
            }
        }

        // Forward checking: Check if enough neighbors are available for remaining vertices
        int remainingVertices = V - pos - 1;
        int availableNeighbors = 0;
        for (int i = 0; i < V; i++) {
            if (graph[v][i] == 1 && !contains(path, i, pos)) {
                availableNeighbors++;
            }
        }
        if (availableNeighbors < remainingVertices) {
            return false; // Not enough neighbors
        }

        return true;
    }

    private static boolean contains(int[] array, int value, int pos) {
        for (int i = 0; i < pos; i++) {
            if (array[i] == value) {
                return true;
            }
        }
        return false;
    }


    // 5. Backtracking with Optimized Adjacency List and Degree Check
    //    - Combines the adjacency list representation with the degree check for better performance.
    public static boolean hasHamiltonianCycle5(List<List<Integer>> adjList) {
        // Degree Check
        for (int i = 0; i < V; i++) {
            if (adjList.get(i).size() < 2) {
                return false;
            }
        }

        int[] path = new int[V];
        Arrays.fill(path, -1);
        path[0] = 0;
        return solveHamiltonianCycle5(adjList, path, 1);
    }

    private static boolean solveHamiltonianCycle5(List<List<Integer>> adjList, int[] path, int pos) {
        if (pos == V) {
            if (adjList.get(path[pos - 1]).contains(0)) {
                return true;
            } else {
                return false;
            }
        }

        for (int v : adjList.get(path[pos - 1])) {
            if (isSafe5(v, path, pos)) {
                path[pos] = v;
                if (solveHamiltonianCycle5(adjList, path, pos + 1)) {
                    return true;
                }
                path[pos] = -1;
            }
        }
        return false;
    }

    private static boolean isSafe5(int v, int[] path, int pos) {
        for (int i = 0; i < pos; i++) {
            if (path[i] == v) {
                return false;
            }
        }
        return true;
    }



    public static void main(String[] args) {
        // Example graph represented as an adjacency matrix
        int[][] graph1 = {
                {0, 1, 0, 1, 0},
                {1, 0, 1, 1, 1},
                {0, 1, 0, 0, 1},
                {1, 1, 0, 0, 1},
                {0, 1, 1, 1, 0},
        };

        // Example graph represented as an adjacency list
        List<List<Integer>> adjList = new ArrayList<>();
        for (int i = 0; i < V; i++) {
            adjList.add(new ArrayList<>());
        }
        adjList.get(0).add(1);
        adjList.get(0).add(3);
        adjList.get(1).add(0);
        adjList.get(1).add(2);
        adjList.get(1).add(3);
        adjList.get(1).add(4);
        adjList.get(2).add(1);
        adjList.get(2).add(4);
        adjList.get(3).add(0);
        adjList.get(3).add(1);
        adjList.get(3).add(4);
        adjList.get(4).add(1);
        adjList.get(4).add(2);
        adjList.get(4).add(3);


        System.out.println("Hamiltonian Cycle Detection:");
        System.out.println("\nUsing Adjacency Matrix:");
        System.out.println("--------------------------");
        System.out.println("Approach 1 (Basic Backtracking): " + hasHamiltonianCycle1(graph1));
        System.out.println("Approach 3 (Degree Check): " + hasHamiltonianCycle3(graph1));
        System.out.println("Approach 4 (Forward Checking): " + hasHamiltonianCycle4(graph1));

        System.out.println("\nUsing Adjacency List:");
        System.out.println("--------------------------");
        System.out.println("Approach 2 (Adjacency List): " + hasHamiltonianCycle2(adjList));
        System.out.println("Approach 5 (Optimized List + Degree Check): " + hasHamiltonianCycle5(adjList));


        // Test case where a Hamiltonian cycle exists
        int[][] graph2 = {
                {0, 1, 0, 0, 1},
                {1, 0, 1, 0, 0},
                {0, 1, 0, 1, 0},
                {0, 0, 1, 0, 1},
                {1, 0, 0, 1, 0},
        };
        System.out.println("\nTest Case with Hamiltonian Cycle:");
        System.out.println("Approach 1 (Basic Backtracking): " + hasHamiltonianCycle1(graph2)); // Should return true


        // Test case where a Hamiltonian cycle does not exist (Degree Check)
        int[][] graph3 = {
                {0, 1, 0, 0, 0},
                {1, 0, 1, 0, 0},
                {0, 1, 0, 0, 0},
                {0, 0, 0, 0, 1},
                {0, 0, 0, 1, 0},
        };
        System.out.println("\nTest Case without Hamiltonian Cycle (Degree < 2):");
        System.out.println("Approach 1 (Basic Backtracking): " + hasHamiltonianCycle1(graph3)); // Should return false, but might take time
        System.out.println("Approach 3 (Degree Check): " + hasHamiltonianCycle3(graph3));       // Should return false quickly

        // Test case with a larger graph (for performance comparison)
        int V_LARGE = 8;
        int[][] graph4 = new int[V_LARGE][V_LARGE];
        for (int i = 0; i < V_LARGE; i++) {
            for (int j = 0; j < V_LARGE; j++) {
                if (Math.random() < 0.3) { // Adjust probability for graph density
                    graph4[i][j] = 1;
                    graph4[j][i] = 1; // Ensure undirected graph
                }
            }
            graph4[i][i] = 0; // No self-loops
        }

        System.out.println("\nPerformance Test (Large Graph, V = " + V_LARGE + "):");
        long startTime = System.nanoTime();
        boolean result1 = hasHamiltonianCycle1(graph4); // Basic backtracking (may be slow)
        long endTime = System.nanoTime();
        System.out.println("Approach 1 (Basic Backtracking): " + result1 + " (Time: " + (endTime - startTime) / 1e6 + " ms)");

        startTime = System.nanoTime();
        boolean result3 = hasHamiltonianCycle3(graph4); // Backtracking with degree check
        endTime = System.nanoTime();
        System.out.println("Approach 3 (Degree Check): " + result3 + " (Time: " + (endTime - startTime) / 1e6 + " ms)");

        startTime = System.nanoTime();
        boolean result4 = hasHamiltonianCycle4(graph4); // Backtracking with forward checking
        endTime = System.nanoTime();
        System.out.println("Approach 4 (Forward Checking): " + result4 + " (Time: " + (endTime - startTime) / 1e6 + " ms)");


    }
}

